### PR TITLE
Support path versioning for rendering responses

### PR DESCRIPTION
**Pull request by GuilhemN**
see https://github.com/FriendsOfSymfony/FOSRestBundle/issues/1604

### CODE REVIEW

File: EventListener/VersionListener.php [hunk 1]
```diff
@@ -53,16 +53,18 @@ public function onKernelRequest(GetResponseEvent $event)
         }
 
         $this->version = $this->versionResolver->resolve($request);
-        if (false === $this->version && null !== $this->defaultVersion) {
+        if (false === $this->version && null !== $this->defaultVersion && !$request->attributes->has('version')) {
             $this->version = $this->defaultVersion;
         }
```
Comment (xabbuh) [inline EventListener/VersionListener.php]: What about changing this to:

```php
if (false === $this->version && $request->attributes->has('version')) {
    $this->version = $request->attributes->get('version');
} elseif (false === $this->version && null !== $this->defaultVersion) {
    $this->version = $this->defaultVersion;
}
```

Comment (xabbuh) [inline EventListener/VersionListener.php]: The code below could then be simplified a bit.

Comment (xabbuh) [inline EventListener/VersionListener.php]: By the way, what was the reason for the `getVersion()` method which requires us to make the event listener stateful?

Comment (GuilhemN) [inline EventListener/VersionListener.php]: In fact we could remove it as we made listeners `@internal`. WDYT?

Comment (xabbuh) [inline EventListener/VersionListener.php]: Oh, I missed the fact that we made it internal in 1.8. So yes, I would update it accordingly. But we can do that in a separate PR.

File: EventListener/VersionListener.php [hunk 2]
```diff
@@ -53,16 +53,24 @@ public function onKernelRequest(GetResponseEvent $event)
...
         $this->version = $this->versionResolver->resolve($request);
+        if (false === $this->version && $request->attributes->has('version')) {
+            $this->version = $request->attributes->get('version');
+        } elseif (false === $this->version && null !== $this->defaultVersion) {
+            $this->version = $this->defaultVersion;
+        }
+
         if (false === $this->version && null !== $this->defaultVersion) {
             $this->version = $this->defaultVersion;
         }
```
Comment (xabbuh) [inline EventListener/VersionListener.php]: This block should be removed now.

File: EventListener/VersionListener.php [hunk 3]
```diff
@@ -53,16 +53,24 @@ public function onKernelRequest(GetResponseEvent $event)
...
+        // Override request attribute
         if (false !== $this->version) {
             $request->attributes->set('version', $this->version);
+        }
 
-            if ($this->viewHandler instanceof ConfigurableViewHandlerInterface) {
-                $this->viewHandler->setExclusionStrategyVersion($this->version);
-            }
+        // Use the resolved version when rendering the response
+        if ($request->attributes->has('version') && $this->viewHandler instanceof ConfigurableViewHandlerInterface) {
```
Comment (xabbuh) [inline EventListener/VersionListener.php]: We can now save a method call here by checking for `$this->version` instead of the request attributes.

File: EventListener/VersionListener.php (hunk 1 above)
Comment (GuilhemN) [inline EventListener/VersionListener.php]: That's alright, it's not much to do and I won't forget it if I do it now :stuck_out_tongue:

**Comment by GuilhemN** [comment]
Do you see something else to change, or can I merge this? :)

**Comment by xabbuh** [comment]
👍 failures do not seem to be related

File: Tests/Functional/VersionTest.php [hunk 4]
```diff
@@ -79,4 +79,16 @@ public function testDefaultVersion()
...
+    public function testVersionInPath()
+    {
+        $this->client->request(
+            'GET',
+            '/version/2.3',
+            [],
+            [],
+            ['HTTP_Accept' => 'application/json']
+        );
+        $this->assertEquals('{"version":"2.3"}', $this->client->getResponse()->getContent());
```
Comment (GuilhemN) [inline Tests/Functional/VersionTest.php]: I agree, will do it.
